lib/libcheri_caprevoke/libcheri_caprevoke.c

```diff
@@ -0,0 +1,438 @@
+/*
+ * Revocation bitmap manipulation operations.
+ */
+
+#include <assert.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stdatomic.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <cheri/cheric.h>
+
+#include <sys/param.h>
+#include <cheri/revoke.h>
+#include <machine/cherireg.h>
+#include <machine/pte.h>
+#include <machine/vmparam.h>
+#include "libcaprevoke.h"
+
+#if defined(__riscv)
```

Comment:
To avoid an explosion of #ifdef's I would suggest doing what we do in libc and libthr and having MACHINE_ARCH subdirectories that get added to -I via CFLAGS in the Makefile and then just having "caprevoke_md.h" in those subdirectories.  The #ifdef for purecap vs hybrid can be pushed into those files and here you would just `#include "caprevoke_md.h"`